### PR TITLE
Add Prometheus metric scrapes 

### PR DESCRIPTION
**Pull request by tgeoghegan**
Adds a config option `PROMETHEUS_CONFIG` that instructs the server to expose a Prometheus metrics scrape endpoint. The value is a JSON document allowing control over the host and port the metrics handler will listen on (which may not be the same host/port as the OHTTP gateway), the path at which scrapes are served and the name of the metric. The only metric currently served is a histogram of response times. Its labeling is rather awkward, in an effort to work within the existing awkward tags on the StatsD metrics, which interpolate HTTP status into result strings.

### CODE REVIEW

**Comment by tgeoghegan** [comment]
I'm not sure if I handled the dependencies correctly, both because I work with Go modules infrequently and I'm not used to this style of having a `go.mod` _and_ vendoring dependencies in the repository. What I did was add the `import` statements in `prometheus_metrics.go` and then ran `go mod vendor` to let the tool sort things out.

<details><summary>Example Prometheus scrape after successfully handling one request and refusing one due to illegal taret</summary>
<p>

```
# HELP go_gc_duration_seconds A summary of the pause duration of garbage collection cycles.
# TYPE go_gc_duration_seconds summary
go_gc_duration_seconds{quantile="0"} 0
go_gc_duration_seconds{quantile="0.25"} 0
go_gc_duration_seconds{quantile="0.5"} 0
go_gc_duration_seconds{quantile="0.75"} 0
go_gc_duration_seconds{quantile="1"} 0
go_gc_duration_seconds_sum 0
go_gc_duration_seconds_count 0
# HELP go_goroutines Number of goroutines that currently exist.
# TYPE go_goroutines gauge
go_goroutines 9
# HELP go_info Information about the Go environment.
# TYPE go_info gauge
go_info{version="go1.20.14"} 1
# HELP go_memstats_alloc_bytes Number of bytes allocated and still in use.
# TYPE go_memstats_alloc_bytes gauge
go_memstats_alloc_bytes 635856
# HELP go_memstats_alloc_bytes_total Total number of bytes allocated, even if freed.
# TYPE go_memstats_alloc_bytes_total counter
go_memstats_alloc_bytes_total 635856
# HELP go_memstats_buck_hash_sys_bytes Number of bytes used by the profiling bucket hash table.
# TYPE go_memstats_buck_hash_sys_bytes gauge
go_memstats_buck_hash_sys_bytes 4227
# HELP go_memstats_frees_total Total number of frees.
# TYPE go_memstats_frees_total counter
go_memstats_frees_total 0
# HELP go_memstats_gc_sys_bytes Number of bytes used for garbage collection system metadata.
# TYPE go_memstats_gc_sys_bytes gauge
go_memstats_gc_sys_bytes 6.727336e+06
# HELP go_memstats_heap_alloc_bytes Number of heap bytes allocated and still in use.
# TYPE go_memstats_heap_alloc_bytes gauge
go_memstats_heap_alloc_bytes 635856
# HELP go_memstats_heap_idle_bytes Number of heap bytes waiting to be used.
# TYPE go_memstats_heap_idle_bytes gauge
go_memstats_heap_idle_bytes 1.163264e+06
# HELP go_memstats_heap_inuse_bytes Number of heap bytes that are in use.
# TYPE go_memstats_heap_inuse_bytes gauge
go_memstats_heap_inuse_bytes 2.572288e+06
# HELP go_memstats_heap_objects Number of allocated objects.
# TYPE go_memstats_heap_objects gauge
go_memstats_heap_objects 1721
# HELP go_memstats_heap_released_bytes Number of heap bytes released to OS.
# TYPE go_memstats_heap_released_bytes gauge
go_memstats_heap_released_bytes 1.163264e+06
# HELP go_memstats_heap_sys_bytes Number of heap bytes obtained from system.
# TYPE go_memstats_heap_sys_bytes gauge
go_memstats_heap_sys_bytes 3.735552e+06
# HELP go_memstats_last_gc_time_seconds Number of seconds since 1970 of last garbage collection.
# TYPE go_memstats_last_gc_time_seconds gauge
go_memstats_last_gc_time_seconds 0
# HELP go_memstats_lookups_total Total number of pointer lookups.
# TYPE go_memstats_lookups_total counter
go_memstats_lookups_total 0
# HELP go_memstats_mallocs_total Total number of mallocs.
# TYPE go_memstats_mallocs_total counter
go_memstats_mallocs_total 1721
# HELP go_memstats_mcache_inuse_bytes Number of bytes in use by mcache structures.
# TYPE go_memstats_mcache_inuse_bytes gauge
go_memstats_mcache_inuse_bytes 28800
# HELP go_memstats_mcache_sys_bytes Number of bytes used for mcache structures obtained from system.
# TYPE go_memstats_mcache_sys_bytes gauge
go_memstats_mcache_sys_bytes 31200
# HELP go_memstats_mspan_inuse_bytes Number of bytes in use by mspan structures.
# TYPE go_memstats_mspan_inuse_bytes gauge
go_memstats_mspan_inuse_bytes 54720
# HELP go_memstats_mspan_sys_bytes Number of bytes used for mspan structures obtained from system.
# TYPE go_memstats_mspan_sys_bytes gauge
go_memstats_mspan_sys_bytes 65280
# HELP go_memstats_next_gc_bytes Number of heap bytes when next garbage collection will take place.
# TYPE go_memstats_next_gc_bytes gauge
go_memstats_next_gc_bytes 4.194304e+06
# HELP go_memstats_other_sys_bytes Number of bytes used for other system allocations.
# TYPE go_memstats_other_sys_bytes gauge
go_memstats_other_sys_bytes 1.166341e+06
# HELP go_memstats_stack_inuse_bytes Number of bytes in use by the stack allocator.
# TYPE go_memstats_stack_inuse_bytes gauge
go_memstats_stack_inuse_bytes 458752
# HELP go_memstats_stack_sys_bytes Number of bytes obtained from system for stack allocator.
# TYPE go_memstats_stack_sys_bytes gauge
go_memstats_stack_sys_bytes 458752
# HELP go_memstats_sys_bytes Number of bytes obtained from system.
# TYPE go_memstats_sys_bytes gauge
go_memstats_sys_bytes 1.2188688e+07
# HELP go_threads Number of OS threads created.
# TYPE go_threads gauge
go_threads 6
# HELP ohttp_gateway_duration 
# TYPE ohttp_gateway_duration histogram
ohttp_gateway_duration_bucket{eventName="gateway_request",method="POST",result="unknown",status="200",le="0.005"} 1
ohttp_gateway_duration_bucket{eventName="gateway_request",method="POST",result="unknown",status="200",le="0.01"} 1
ohttp_gateway_duration_bucket{eventName="gateway_request",method="POST",result="unknown",status="200",le="0.025"} 1
ohttp_gateway_duration_bucket{eventName="gateway_request",method="POST",result="unknown",status="200",le="0.05"} 1
ohttp_gateway_duration_bucket{eventName="gateway_request",method="POST",result="unknown",status="200",le="0.1"} 1
ohttp_gateway_duration_bucket{eventName="gateway_request",method="POST",result="unknown",status="200",le="0.25"} 1
ohttp_gateway_duration_bucket{eventName="gateway_request",method="POST",result="unknown",status="200",le="0.5"} 1
ohttp_gateway_duration_bucket{eventName="gateway_request",method="POST",result="unknown",status="200",le="1"} 2
ohttp_gateway_duration_bucket{eventName="gateway_request",method="POST",result="unknown",status="200",le="2.5"} 2
ohttp_gateway_duration_bucket{eventName="gateway_request",method="POST",result="unknown",status="200",le="5"} 2
ohttp_gateway_duration_bucket{eventName="gateway_request",method="POST",result="unknown",status="200",le="10"} 2
ohttp_gateway_duration_bucket{eventName="gateway_request",method="POST",result="unknown",status="200",le="+Inf"} 2
ohttp_gateway_duration_sum{eventName="gateway_request",method="POST",result="unknown",status="200"} 1
ohttp_gateway_duration_count{eventName="gateway_request",method="POST",result="unknown",status="200"} 2
ohttp_gateway_duration_bucket{eventName="gateway_request",method="unknown",result="gateway_payload200",status="unknown",le="0.005"} 0
ohttp_gateway_duration_bucket{eventName="gateway_request",method="unknown",result="gateway_payload200",status="unknown",le="0.01"} 0
ohttp_gateway_duration_bucket{eventName="gateway_request",method="unknown",result="gateway_payload200",status="unknown",le="0.025"} 0
ohttp_gateway_duration_bucket{eventName="gateway_request",method="unknown",result="gateway_payload200",status="unknown",le="0.05"} 0
ohttp_gateway_duration_bucket{eventName="gateway_request",method="unknown",result="gateway_payload200",status="unknown",le="0.1"} 0
ohttp_gateway_duration_bucket{eventName="gateway_request",method="unknown",result="gateway_payload200",status="unknown",le="0.25"} 0
ohttp_gateway_duration_bucket{eventName="gateway_request",method="unknown",result="gateway_payload200",status="unknown",le="0.5"} 0
ohttp_gateway_duration_bucket{eventName="gateway_request",method="unknown",result="gateway_payload200",status="unknown",le="1"} 1
ohttp_gateway_duration_bucket{eventName="gateway_request",method="unknown",result="gateway_payload200",status="unknown",le="2.5"} 1
ohttp_gateway_duration_bucket{eventName="gateway_request",method="unknown",result="gateway_payload200",status="unknown",le="5"} 1
ohttp_gateway_duration_bucket{eventName="gateway_request",method="unknown",result="gateway_payload200",status="unknown",le="10"} 1
ohttp_gateway_duration_bucket{eventName="gateway_request",method="unknown",result="gateway_payload200",status="unknown",le="+Inf"} 1
ohttp_gateway_duration_sum{eventName="gateway_request",method="unknown",result="gateway_payload200",status="unknown"} 1
ohttp_gateway_duration_count{eventName="gateway_request",method="unknown",result="gateway_payload200",status="unknown"} 1
ohttp_gateway_duration_bucket{eventName="gateway_request",method="unknown",result="gateway_payload403",status="unknown",le="0.005"} 1
ohttp_gateway_duration_bucket{eventName="gateway_request",method="unknown",result="gateway_payload403",status="unknown",le="0.01"} 1
ohttp_gateway_duration_bucket{eventName="gateway_request",method="unknown",result="gateway_payload403",status="unknown",le="0.025"} 1
ohttp_gateway_duration_bucket{eventName="gateway_request",method="unknown",result="gateway_payload403",status="unknown",le="0.05"} 1
ohttp_gateway_duration_bucket{eventName="gateway_request",method="unknown",result="gateway_payload403",status="unknown",le="0.1"} 1
ohttp_gateway_duration_bucket{eventName="gateway_request",method="unknown",result="gateway_payload403",status="unknown",le="0.25"} 1
ohttp_gateway_duration_bucket{eventName="gateway_request",method="unknown",result="gateway_payload403",status="unknown",le="0.5"} 1
ohttp_gateway_duration_bucket{eventName="gateway_request",method="unknown",result="gateway_payload403",status="unknown",le="1"} 1
ohttp_gateway_duration_bucket{eventName="gateway_request",method="unknown",result="gateway_payload403",status="unknown",le="2.5"} 1
ohttp_gateway_duration_bucket{eventName="gateway_request",method="unknown",result="gateway_payload403",status="unknown",le="5"} 1
ohttp_gateway_duration_bucket{eventName="gateway_request",method="unknown",result="gateway_payload403",status="unknown",le="10"} 1
ohttp_gateway_duration_bucket{eventName="gateway_request",method="unknown",result="gateway_payload403",status="unknown",le="+Inf"} 1
ohttp_gateway_duration_sum{eventName="gateway_request",method="unknown",result="gateway_payload403",status="unknown"} 0
ohttp_gateway_duration_count{eventName="gateway_request",method="unknown",result="gateway_payload403",status="unknown"} 1
ohttp_gateway_duration_bucket{eventName="gateway_request",method="unknown",result="request_forbidden",status="unknown",le="0.005"} 1
ohttp_gateway_duration_bucket{eventName="gateway_request",method="unknown",result="request_forbidden",status="unknown",le="0.01"} 1
ohttp_gateway_duration_bucket{eventName="gateway_request",method="unknown",result="request_forbidden",status="unknown",le="0.025"} 1
ohttp_gateway_duration_bucket{eventName="gateway_request",method="unknown",result="request_forbidden",status="unknown",le="0.05"} 1
ohttp_gateway_duration_bucket{eventName="gateway_request",method="unknown",result="request_forbidden",status="unknown",le="0.1"} 1
ohttp_gateway_duration_bucket{eventName="gateway_request",method="unknown",result="request_forbidden",status="unknown",le="0.25"} 1
ohttp_gateway_duration_bucket{eventName="gateway_request",method="unknown",result="request_forbidden",status="unknown",le="0.5"} 1
ohttp_gateway_duration_bucket{eventName="gateway_request",method="unknown",result="request_forbidden",status="unknown",le="1"} 1
ohttp_gateway_duration_bucket{eventName="gateway_request",method="unknown",result="request_forbidden",status="unknown",le="2.5"} 1
ohttp_gateway_duration_bucket{eventName="gateway_request",method="unknown",result="request_forbidden",status="unknown",le="5"} 1
ohttp_gateway_duration_bucket{eventName="gateway_request",method="unknown",result="request_forbidden",status="unknown",le="10"} 1
ohttp_gateway_duration_bucket{eventName="gateway_request",method="unknown",result="request_forbidden",status="unknown",le="+Inf"} 1
ohttp_gateway_duration_sum{eventName="gateway_request",method="unknown",result="request_forbidden",status="unknown"} 0
ohttp_gateway_duration_count{eventName="gateway_request",method="unknown",result="request_forbidden",status="unknown"} 1
ohttp_gateway_duration_bucket{eventName="gateway_request",method="unknown",result="success",status="unknown",le="0.005"} 0
ohttp_gateway_duration_bucket{eventName="gateway_request",method="unknown",result="success",status="unknown",le="0.01"} 0
ohttp_gateway_duration_bucket{eventName="gateway_request",method="unknown",result="success",status="unknown",le="0.025"} 0
ohttp_gateway_duration_bucket{eventName="gateway_request",method="unknown",result="success",status="unknown",le="0.05"} 0
ohttp_gateway_duration_bucket{eventName="gateway_request",method="unknown",result="success",status="unknown",le="0.1"} 0
ohttp_gateway_duration_bucket{eventName="gateway_request",method="unknown",result="success",status="unknown",le="0.25"} 0
ohttp_gateway_duration_bucket{eventName="gateway_request",method="unknown",result="success",status="unknown",le="0.5"} 0
ohttp_gateway_duration_bucket{eventName="gateway_request",method="unknown",result="success",status="unknown",le="1"} 1
ohttp_gateway_duration_bucket{eventName="gateway_request",method="unknown",result="success",status="unknown",le="2.5"} 1
ohttp_gateway_duration_bucket{eventName="gateway_request",method="unknown",result="success",status="unknown",le="5"} 1
ohttp_gateway_duration_bucket{eventName="gateway_request",method="unknown",result="success",status="unknown",le="10"} 1
ohttp_gateway_duration_bucket{eventName="gateway_request",method="unknown",result="success",status="unknown",le="+Inf"} 1
ohttp_gateway_duration_sum{eventName="gateway_request",method="unknown",result="success",status="unknown"} 1
ohttp_gateway_duration_count{eventName="gateway_request",method="unknown",result="success",status="unknown"} 1
# HELP process_cpu_seconds_total Total user and system CPU time spent in seconds.
# TYPE process_cpu_seconds_total counter
process_cpu_seconds_total 0.01
# HELP process_max_fds Maximum number of open file descriptors.
# TYPE process_max_fds gauge
process_max_fds 524288
# HELP process_open_fds Number of open file descriptors.
# TYPE process_open_fds gauge
process_open_fds 10
# HELP process_resident_memory_bytes Resident memory size in bytes.
# TYPE process_resident_memory_bytes gauge
process_resident_memory_bytes 9.437184e+06
# HELP process_start_time_seconds Start time of the process since unix epoch in seconds.
# TYPE process_start_time_seconds gauge
process_start_time_seconds 1.71624443907e+09
# HELP process_virtual_memory_bytes Virtual memory size in bytes.
# TYPE process_virtual_memory_bytes gauge
process_virtual_memory_bytes 7.36333824e+08
# HELP process_virtual_memory_max_bytes Maximum amount of virtual memory available in bytes.
# TYPE process_virtual_memory_max_bytes gauge
process_virtual_memory_max_bytes 1.8446744073709552e+19
# HELP promhttp_metric_handler_requests_in_flight Current number of scrapes being served.
# TYPE promhttp_metric_handler_requests_in_flight gauge
promhttp_metric_handler_requests_in_flight 1
# HELP promhttp_metric_handler_requests_total Total number of scrapes by HTTP status code.
# TYPE promhttp_metric_handler_requests_total counter
promhttp_metric_handler_requests_total{code="200"} 1
promhttp_metric_handler_requests_total{code="500"} 0
promhttp_metric_handler_requests_total{code="503"} 0
```

</p>
</details>

File: main.go
Comment (tgeoghegan) [inline main.go:283]: There was a bug here where we used to use `strconv.ParseInt(os.Getenv(statsdTimeoutVariable))`, and if one wasn't using StatsD metrics, you'd always get an annoying "Failed parsing metrics timeout:" logline. Now we use `getStringEnv` and default to 100, which matches the existing behavior without nagging operators.

**Review by bwesterb — APPROVED** [review]

**Comment by bwesterb** [comment]
Thanks for this. Needs second code owner review.

File: prometheus_metrics.go
Comment (thibmeu) [inline prometheus_metrics.go:52]: You want to use [`net.JoinHostPort`](https://pkg.go.dev/net#JoinHostPort)

```suggestion
		Addr:    net.JoinHostPort(config.Host, config.Port),
```

File: prometheus_metrics.go
Comment (thibmeu) [inline prometheus_metrics.go:58]: nits: it might be interesting to provide a context to the factory, so the server can be killed independently of the whole process.

**Review by thibmeu — APPROVED** [review]
Looks good to me. Only minor comments.

**Comment by bwesterb** [comment]
Let me know when you want me to press the merge button.

File: prometheus_metrics.go
Comment (tgeoghegan) [reply prometheus_metrics.go:52]: Sure. Do you also want to fix this in [`statsd_metrics.go`](https://github.com/cloudflare/privacy-gateway-server-go/blob/1d8081f775055dd2f078e81da52fa517608f4a58/statsd_metrics.go#L37)?

File: prometheus_metrics.go
Comment (tgeoghegan) [reply prometheus_metrics.go:58]: Interesting idea! Would this entail changes to the `MetricsFactory` interface in `metrics.go` so that there's a generic way to "stop" metrics?

**Comment by tgeoghegan** [comment]
@bwesterb I squashed the PR branch and force-pushed. IMO it's good to go, assuming the tests still pass. Thanks!

File: prometheus_metrics.go
Comment (thibmeu) [reply prometheus_metrics.go:67]: I'd say adding `Stop() error` is a way to do this, or enforcing passing a `Context`, which can then be cancelled if necessary.
It's beyond the scope of this change though.

**Comment by thibmeu** [comment]
@bwesterb I'll let you press the merge button. All good for me.

**Comment by bwesterb** [comment]
Thank you @tgeoghegan !

**Pull request merged**
